src/stability_sdk/api.py

```diff
@@ -0,0 +1,662 @@
+import grpc
+import io
+import logging
+import random
+import time
+
+from google.protobuf.struct_pb2 import Struct
+from PIL import Image
+from typing import Any, Dict, List, Optional, Sequence, Tuple, Union
+
+import stability_sdk.interfaces.gooseai.dashboard.dashboard_pb2 as dashboard
+import stability_sdk.interfaces.gooseai.dashboard.dashboard_pb2_grpc as dashboard_grpc
+import stability_sdk.interfaces.gooseai.generation.generation_pb2 as generation
+import stability_sdk.interfaces.gooseai.generation.generation_pb2_grpc as generation_grpc
+
+from .utils import (
+    image_mix,
+    image_to_prompt,
+    tensor_to_prompt,
+)
+
+
+logger = logging.getLogger(__name__)
+logger.setLevel(level=logging.INFO)
+
+
+def open_channel(host: str, api_key: str = None, max_message_len: int = 20*1024*1024) -> grpc.Channel:
+    options=[
+        ('grpc.max_send_message_length', max_message_len),
+        ('grpc.max_receive_message_length', max_message_len),
+    ]    
+    if host.endswith(":443"):
+        call_credentials = [grpc.access_token_call_credentials(api_key)]
+        channel_credentials = grpc.composite_channel_credentials(
+            grpc.ssl_channel_credentials(), *call_credentials
+        )
+        channel = grpc.secure_channel(host, channel_credentials, options=options)
+    else:
+        channel = grpc.insecure_channel(host, options=options)
+    return channel
+
+
+class ClassifierException(Exception):
+    """Raised when server classifies generated content as inappropriate.
+
+    Attributes:
+        classifier_result: Categories the result image exceeded the threshold for
+        prompt: The prompt that was classified as inappropriate
+    """    
+    def __init__(self, classifier_result: Optional[generation.ClassifierParameters]=None, prompt: Optional[str]=None):
+        self.classifier_result = classifier_result
+        self.prompt = prompt
+
+class OutOfCreditsException(Exception):
+    """Raised when account doesn't have enough credits to perform a request."""
+    def __init__(self, details: str):
+        self.details = details
+
+
+class Endpoint:
+    def __init__(self, stub, engine_id):
+        self.stub = stub
+        self.engine_id = engine_id
+
+
+class Context:
+    def __init__(
+            self, 
+            host: str="", 
+            api_key: str=None, 
+            stub: generation_grpc.GenerationServiceStub=None,
+            generate_engine_id: str="stable-diffusion-xl-beta-v2-2-2",
```

Comment:
Should either pull this value from `animation.DEFAULT_MODEL`, or the animation submodule so infer that variables value from the` Context.generate_engine_id` default